File: tools/progress_tracker.py
#!/usr/bin/env python3
"""Lightweight, read-only progress monitor for active Glimpse ABM runs."""

from __future__ import annotations

import argparse
import re
import sys
import time
from dataclasses import dataclass
from pathlib import Path
from typing import Dict, Iterable, Optional

import pandas as pd

RUN_DIR_PATTERN = re.compile(r"emergent_run_(\d+)")


@dataclass(frozen=True)
class RunProgress:
    name: str
    rounds_completed: int

    @property
    def round_label(self) -> str:
        return f"{self.rounds_completed:>5d}"


def _sorted_run_dirs(results_dir: Path) -> Iterable[Path]:
    for child in sorted(results_dir.iterdir()):
        if not child.is_dir():
            continue
        match = RUN_DIR_PATTERN.fullmatch(child.name)
        if match:
            yield child


def _load_latest_round_from_subdir(subdir: Path) -> Optional[int]:
    """Return zero-based round index from the newest .pkl, if available."""
    if not subdir.exists():
        return None

    files = sorted(subdir.glob("*.pkl"))
    for path in reversed(files):
        try:
            df = pd.read_pickle(path)
        except (FileNotFoundError, pd.errors.EmptyDataError, EOFError, ValueError):
            continue
        if "round" not in df.columns:
            continue
        rounds = (
            pd.to_numeric(df["round"], errors="coerce")
            .dropna()
            .astype(int)
        )
        if not rounds.empty:
            return int(rounds.max())
    return None


def _load_latest_round_from_log(log_path: Path) -> Optional[int]:
    """Return zero-based round index by counting entries in run_log.jsonl."""
    if not log_path.exists():
        return None
    try:
        with log_path.open("r", encoding="utf-8") as handle:
            count = sum(1 for line in handle if line.strip())
        return count - 1 if count > 0 else None
    except OSError:
        return None


def _detect_run_progress(run_dir: Path) -> RunProgress:
    candidates = [
        _load_latest_round_from_log(run_dir / "run_log.jsonl"),
        _load_latest_round_from_subdir(run_dir / "summary"),
        _load_latest_round_from_subdir(run_dir / "decisions"),
        _load_latest_round_from_subdir(run_dir / "market"),
    ]
    max_round = max((c for c in candidates if c is not None), default=-1)
    rounds_completed = max_round + 1 if max_round >= 0 else 0
    return RunProgress(name=run_dir.name, rounds_completed=rounds_completed)


def collect_progress(results_dir: Path, limit: Optional[int] = None) -> Dict[str, RunProgress]:
    progress: Dict[str, RunProgress] = {}
    for idx, run_dir in enumerate(_sorted_run_dirs(results_dir)):
        if limit is not None and idx >= limit:
            break
        progress[run_dir.name] = _detect_run_progress(run_dir)
    return progress


def render_report(progress: Dict[str, RunProgress]) -> str:
    if not progress:
        return "No run directories discovered yet."

    ordered = sorted(progress.values(), key=lambda rp: int(RUN_DIR_PATTERN.match(rp.name).group(1)))
    total_rounds = sum(rp.rounds_completed for rp in ordered)
    max_rounds = max((rp.rounds_completed for rp in ordered), default=0)
    active_runs = sum(1 for rp in ordered if rp.rounds_completed > 0)

    lines = [
        f"Runs tracked: {len(ordered)}  |  Active: {active_runs}  |  Total rounds: {total_rounds}  |  Max rounds: {max_rounds}",
        "Recent runs:",
    ]

    preview = ordered[-5:] if len(ordered) > 5 else ordered
    for rp in preview:
        lines.append(f"  {rp.name:<16} -> rounds: {rp.round_label}")
    return "\n".join(lines)


def main() -> None:
    parser = argparse.ArgumentParser(description="Monitor Glimpse ABM run progress without touching active writers.")
    parser.add_argument("results_dir", type=Path, help="Path to the simulation results directory.")
    parser.add_argument(
        "--limit",
        type=int,
        default=None,
        help="Only consider the first N run directories (sorted by index).",
    )
    parser.add_argument(
        "--interval",
        type=float,
        default=15.0,
        help="Polling interval in seconds for continuous monitoring.",
    )
    parser.add_argument(
        "--once",
        action="store_true",
        help="Collect a single snapshot and exit.",
    )
    args = parser.parse_args()

    if not args.results_dir.exists():
        print(f"Results directory not found: {args.results_dir}", file=sys.stderr)
        sys.exit(1)

    while True:
        progress = collect_progress(args.results_dir, limit=args.limit)
        timestamp = time.strftime("%Y-%m-%d %H:%M:%S")
        print(f"[{timestamp}]")
        print(render_report(progress))
        if args.once:
            break
        time.sleep(args.interval)


if __name__ == "__main__":
    main()

File: julia/tools/progress_tracker.jl
"""
Lightweight, read-only progress monitor for active Glimpse ABM runs.

Port of: glimpse_abm/tools/progress_tracker.py
"""

module ProgressTracker

using Dates
using DataFrames
using Arrow

export RunProgress, collect_progress, render_report, watch_progress

const RUN_DIR_PATTERN = r"emergent_run_(\d+)"

"""
Progress information for a single run.
"""
struct RunProgress
    name::String
    rounds_completed::Int
end

"""
Get formatted round label.
"""
function round_label(rp::RunProgress)::String
    return lpad(string(rp.rounds_completed), 5)
end

"""
Sort run directories by index.
"""
function sorted_run_dirs(results_dir::String)::Vector{String}
    dirs = String[]

    if !isdir(results_dir)
        return dirs
    end

    for entry in readdir(results_dir)
        path = joinpath(results_dir, entry)
        if isdir(path) && occursin(RUN_DIR_PATTERN, entry)
            push!(dirs, path)
        end
    end

    # Sort by run index
    sort!(dirs, by=d -> begin
        m = match(RUN_DIR_PATTERN, basename(d))
        isnothing(m) ? 0 : parse(Int, m.captures[1])
    end)

    return dirs
end

"""
Load latest round from Arrow files in a subdirectory.
"""
function load_latest_round_from_subdir(subdir::String)::Union{Int,Nothing}
    if !isdir(subdir)
        return nothing
    end

    # Check for Arrow files
    arrow_files = filter(f -> endswith(f, ".arrow"), readdir(subdir))
    if isempty(arrow_files)
        return nothing
    end

    # Sort by name (assuming batch_N.arrow format)
    sort!(arrow_files, by=f -> begin
        m = match(r"batch_(\d+)", f)
        isnothing(m) ? 0 : parse(Int, m.captures[1])
    end, rev=true)

    for file in arrow_files
        try
            path = joinpath(subdir, file)
            df = DataFrame(Arrow.Table(path))
            if "round" in names(df) && nrow(df) > 0
                rounds = df.round
                valid_rounds = filter(x -> !ismissing(x), rounds)
                if !isempty(valid_rounds)
                    return Int(maximum(valid_rounds))
                end
            end
        catch
            continue
        end
    end

    return nothing
end

"""
Load latest round from run log file.
"""
function load_latest_round_from_log(log_path::String)::Union{Int,Nothing}
    if !isfile(log_path)
        return nothing
    end

    try
        count = 0
        open(log_path, "r") do f
            for line in eachline(f)
                if !isempty(strip(line))
                    count += 1
                end
            end
        end
        return count > 0 ? count - 1 : nothing
    catch
        return nothing
    end
end

"""
Detect progress for a single run directory.
"""
function detect_run_progress(run_dir::String)::RunProgress
    candidates = [
        load_latest_round_from_log(joinpath(run_dir, "run_log.jsonl")),
        load_latest_round_from_subdir(joinpath(run_dir, "summary")),
        load_latest_round_from_subdir(joinpath(run_dir, "decisions")),
        load_latest_round_from_subdir(joinpath(run_dir, "market"))
    ]

    # Filter out nothing values and get max
    valid = filter(!isnothing, candidates)
    max_round = isempty(valid) ? -1 : maximum(valid)
    rounds_completed = max_round >= 0 ? max_round + 1 : 0

    return RunProgress(basename(run_dir), rounds_completed)
end

"""
Collect progress from all run directories.
"""
function collect_progress(results_dir::String; limit::Union{Int,Nothing}=nothing)::Dict{String,RunProgress}
    progress = Dict{String,RunProgress}()

    run_dirs = sorted_run_dirs(results_dir)

    for (idx, run_dir) in enumerate(run_dirs)
        if !isnothing(limit) && idx > limit
            break
        end
        name = basename(run_dir)
        progress[name] = detect_run_progress(run_dir)
    end

    return progress
end

"""
Render a progress report as a string.
"""
function render_report(progress::Dict{String,RunProgress})::String
    if isempty(progress)
        return "No run directories discovered yet."
    end

    # Sort by run index
    ordered = sort(collect(values(progress)), by=rp -> begin
        m = match(RUN_DIR_PATTERN, rp.name)
        isnothing(m) ? 0 : parse(Int, m.captures[1])
    end)

    total_rounds = sum(rp.rounds_completed for rp in ordered)
    max_rounds = maximum(rp.rounds_completed for rp in ordered; init=0)
    active_runs = count(rp -> rp.rounds_completed > 0, ordered)

    lines = [
        "Runs tracked: $(length(ordered))  |  Active: $active_runs  |  Total rounds: $total_rounds  |  Max rounds: $max_rounds",
        "Recent runs:"
    ]

    # Show last 5 runs
    preview = length(ordered) > 5 ? ordered[end-4:end] : ordered
    for rp in preview
        push!(lines, "  $(rpad(rp.name, 16)) -> rounds: $(round_label(rp))")
    end

    return join(lines, "\n")
end

"""
Watch progress continuously with polling interval.
"""
function watch_progress(
    results_dir::String;
    interval::Float64=15.0,
    limit::Union{Int,Nothing}=nothing,
    once::Bool=false
)
    if !isdir(results_dir)
        println("Results directory not found: $results_dir")
        return
    end

    while true
        progress = collect_progress(results_dir; limit=limit)
        timestamp = Dates.format(now(), "yyyy-mm-dd HH:MM:SS")
        println("[$timestamp]")
        println(render_report(progress))

        if once
            break
        end

        sleep(interval)
    end
end

"""
Main entry point for command-line usage.
"""
function main(args=ARGS)
    if isempty(args)
        println("Usage: julia progress_tracker.jl <results_dir> [--once] [--interval=N] [--limit=N]")
        return
    end

    results_dir = args[1]
    once = "--once" in args

    interval = 15.0
    for arg in args
        if startswith(arg, "--interval=")
            interval = parse(Float64, split(arg, "=")[2])
        end
    end

    limit = nothing
    for arg in args
        if startswith(arg, "--limit=")
            limit = parse(Int, split(arg, "=")[2])
        end
    end

    watch_progress(results_dir; interval=interval, limit=limit, once=once)
end

end # module ProgressTracker
